### ISSUE TITLE
Integration with Github as a data portal

### ISSUE THREAD

**Issue by roll** [frictionlessdata/frictionless-py#1185]
# Overview

An important step for Frictionless Framework is to provide an ability to read and write packages from different data portals (CKAN/Github/Zenodod/etc) so the users can publish and access their packages easily and using a straightforward API. This issue is for Github integration. The implementation is already prototyped in v5 branch.

### Read package

Read package from a repository that has a datapackage.json/yaml:

```python
package = Package("https://github.com/datasets/population")
package = Package.from_github(...) # alias
```

Read package from a repository without a datapackage.json/yaml. We probably need to filter files and add only CSV/XLS(X) to the package. Also GithubControl should have this configurable. We need to map as much as possible metadata provided for Github Repository:

```python
package = Package("https://github.com/frictionlessdata/repository-demo")
package = Package("<link>", control=portals.GithubControl(formats=['csv']))
```

### Write package

Publish a package on Github (for now, only if the repo doesn't exist). Also we need to provide an ability to store credentials in ENV/etc. We need to map as much as possible metadata provided by Package:

```python
package.to_github(user=, repo=, api_key=)
```

### Read catalog

Read catalog from github search. Design some search configurations like `limit` and `offset` (pagination).

```python
catalog = Catalog(control=portals.GithubControl(search="<frictionless>")
for package in catalog.packages:
  print(package.name)
```

## Plan

- [ ] prototype the functionality based on the functional requirements
- [ ] get feedback from @roll on the implementation
- [ ] finish the implementation
- [ ] design the testing approach (probably using pytest.vcr fro reading but how to test writing?)
- [ ] write a great deal of tests to be sure that the integration works correctly
- [ ] write a comprehensive tutorial - https://framework.frictionlessdata.io/docs/tutorials/tutorials-overview (new section Portals Tutorials)

**Comment by shashigharti** [comment]
@roll, this is what I prototyped for the github api. I am still adding the tests. WDYT?

For "write" testing  pytest.mark.vcr works fine. I checked the code of PyGithub and it is also using data replay
https://github.com/PyGithub/PyGithub/tree/master/tests/ReplayData

```
import os
from collections import namedtuple
from typing import Optional
from .control import GithubControl
from ...exception import FrictionlessException
from ...package import Package
from ...package import Manager
from ...platform import platform
from ...resource import Resource


class GithubManager(Manager[GithubControl]):

    # Read

    def read_catalog(self, user: Optional[str] = None):
        client = platform.github.Github(per_page=self.control.per_page)
        user = user or self.control.user
        search = self.control.search or f"user:{user}"
        assert search
        if self.control.page:
            search = f"{search}&page={self.control.page}"

        # Search
        try:
            repositories = client.search_repositories(search)
        except Exception as exception:
            note = "Github API error" + repr(exception)
            raise FrictionlessException(note)
        packages = []
        for repository in repositories:
            branch = repository.default_branch
            base_path = (
                f"https://raw.githubusercontent.com/{repository.full_name}/{branch}"
            )
            contents = repository.get_contents("")
            paths = _get_contents(contents, repository)
            package = _get_package(paths, repository, base_path, self.control.formats)

            # return packages which has resources
            if package.resources:
                packages.append(package)
        if not packages:
            return packages
        note = "Package not found"
        raise FrictionlessException(note)

    def read_package(self, *, user: Optional[str] = None, repo: Optional[str] = None):
        client = platform.github.Github()
        user = user or self.control.user
        repo = repo or self.control.repo
        assert user
        assert repo
        location = "/".join([user, repo])
        try:
            repository = client.get_repo(location)
        except Exception as exception:
            note = "Github API error" + repr(exception)
            raise FrictionlessException(note)
        branch = repository.default_branch
        base_path = f"https://raw.githubusercontent.com/{location}/{branch}"
        contents = repository.get_contents("")
        paths = _get_contents(contents, repository)
        package = _get_package(paths, repo, base_path, self.control.formats)
        if not package.resources:
            return package
        note = "Package not found"
        raise FrictionlessException(note)

    # Write

    def write_package(
        self,
        *,
        repo: Optional[str] = None,
        apikey: Optional[str] = None,
        filename: Optional[str] = None,
        content: Optional[str] = None,
        branch: Optional[str] = None,
    ):
        repo = repo or self.control.repo
        apikey = apikey or os.environ.get("GITHUB_ACCESS_TOKEN", None)
        assert repo
        assert apikey
        # Create repo
        repository = None
        user = None
        try:
            client = platform.github.Github(apikey)
            user = client.get_user()
            repository = user.create_repo(
                name=repo, auto_init=True, gitignore_template="Python"
            )
        except Exception as exception:
            note = "Github API error" + repr(exception)
            raise FrictionlessException(note)

        # Write files to the repo
        file_created = None
        if content:
            try:
                newfile = filename or "package.json"
                repository = user.get_repo(repo)
                author = platform.github.InputGitAuthor(user.name, user.email)
                branch = branch or repository.default_branch
                file_created = repository.create_file(
                    path=newfile,
                    message="Create package.json",
                    content=content,
                    branch=branch,
                    committer=author,
                    author=author,
                )
            except Exception as exception:
                note = "Github API error" + repr(exception)
                raise FrictionlessException(note)

        response = namedtuple("response", ["repository", "file_created"])
        return response(repository, file_created)


def _get_contents(contents, repository):
    paths = []
    while contents:
        file_content = contents.pop(0)  # type: ignore
        if file_content.type == "dir":
            contents.extend(repository.get_contents(file_content.path))  # type: ignore
        else:
            paths.append(file_content.path)
    return paths


def _get_package(paths, repository, base_path, formats):
    package = Package(name=repository.name)
    for path in paths:
        fullpath = f"{base_path}/{path}"
        if path in ["datapackage.json", "datapackage.yaml"]:
            return Package.from_descriptor(fullpath)
        if any(path.endswith(ext) for ext in formats):
            package.add_resource(Resource(path=fullpath))
    return package

```

**Comment by roll** [comment]
@shashigharti 
Can you please also implement publishing functions on the Package class?

```
package.publish(self, target: Optional[Any] = None, *, control: Optional[Control] = None) -> None
package.to_github(self, target, *, control)
package.from_github(self, target, *, control)
```

**Comment by roll** [comment]
Someting like this:

```
def publish(self, target: Optional[Any] = None, *, control: Optional[Control] = None) -> None:
            manager = system.create_manager(source, control=control)
            if manager:
                package = manager.write_package(self)
            # raise: unsupported target
```

Then we need to use it in tests for writing/publishing

**Comment by shashigharti** [comment]
Thanks. I have implemented to_github and from_github. I will add publish also.

**Comment by shashigharti** [comment]
@roll I have made changes as per your feedback and added new tests. It is ready for review.
Few questions I have in mind:

- I added publish function but both to_github and publish has same functionality 'write_package' should I add new function for publish in manager.
- Write function only writes datapackage.json file. Shall I add function to write resources?
- I haven't yet added tutorial, I will tutorial once the functionality is completed and approved.
